### PR TITLE
chore: Make KeyIndex iterable

### PR DESCRIPTION
**Pull request by dranikpg**
* Fixes a bug in cluster slot checks where we miss the bonus key
* Makes KeyIndex iterable, so we don't have to loop manually, avoiding bugs like this one

### CODE REVIEW

File: src/server/tx_base.h [hunk 1]
```diff
@@ -34,28 +35,50 @@ struct KeyLockArgs {
...
-  std::optional<uint16_t> bonus{};
+  unsigned operator*() {
+    if (bonus)
+      return *bonus;
+    return start;
+  }
 
-  KeyIndex(unsigned s = 0, unsigned e = 0, unsigned step = 0) : start(s), end(e), step(step) {
+  KeyIndex& operator++() {
+    if (bonus)
```
Comment (romange) [inline src/server/tx_base.h:56]: it's a bit dangerous and counter intuitive

Comment (dranikpg) [reply src/server/tx_base.h:56]: Then suggest a more elegant way 🙂 I thinks it's still better than writing `if (bonus)` in high level logic

**Comment by adiholden** [comment]
how did you found this bug? do we have a test covering this?

**Comment by dranikpg** [comment]
> how did you found this bug?

Just by reading the code

> do we have a test covering this?

No, but because it's using the iterator now, it's validitiy is tested by all other usages as well

**Comment by dranikpg** [comment]
Added a test

File: src/server/transaction.cc [hunk 2]
```diff
@@ -190,32 +190,22 @@ void Transaction::InitGlobal() {
...
+  DCHECK(!key_index.bonus || key_index.step == 1);
+
   auto& shard_index = *out;
+  for (unsigned i : key_index.Range()) {
+    string_view key = ArgS(full_args_, i);
+    unique_slot_checker_.Add(key);
+    ShardId sid = Shard(key, shard_data_.size());
 
-  auto add = [&shard_index](uint32_t sid, uint32_t b, uint32_t e) {
+    unsigned step = shard_index[sid].key_step = key_index.bonus ? 1 : key_index.step;
```
Comment (romange) [inline src/server/transaction.cc:202]: is it `=` or `==` here?

File: src/server/transaction.cc [hunk 3]
```diff
@@ -295,14 +281,14 @@ void Transaction::InitByKeys(const KeyIndex& key_index) {
   // Stub transactions always operate only on single shard.
   bool is_stub = multi_ && multi_->role == SQUASHED_STUB;
 
-  if ((key_index.HasSingleKey() && !IsAtomicMulti()) || is_stub) {
+  if ((key_index.NumArgs() == 1 && !IsAtomicMulti()) || is_stub) {
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/transaction.cc:284]: I wanted to do it first, but it's also used as literally num args. And a single argument can only exist with a single key in that context

File: src/server/transaction.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/transaction.cc:202]: it's =, I just re-use the variable step below, while also storing in at shard_index[sid].key_step

Comment (romange) [reply src/server/transaction.cc:202]: please separate this.

**Review by romange — APPROVED** [review]

**Pull request merged**